src/core/electrostatics/p3m.cpp

```diff
@@ -366,24 +371,28 @@ template <int cao> struct AssignForces {
                                        p3m.E_mesh[2][ind]};
         });
 
-        p.force() -= pref * force;
+        p_force -= pref * force;
         ++p_index;
       }
     }
   }
 };
 
-auto dipole_moment(Particle const &p, BoxGeometry const &box) {
-  return p.q() * unfolded_position(p.pos(), p.image_box(), box.length());
+auto dipole_moment(double const q, Utils::Vector3d const unfolded_pos) {
+  return q * unfolded_pos;
 }
 
+template <typename combined_ranges>
 auto calc_dipole_moment(boost::mpi::communicator const &comm,
-                        ParticleRange const &particles,
-                        BoxGeometry const &box) {
-  auto const local_dip = boost::accumulate(
-      particles, Utils::Vector3d{}, [&box](Utils::Vector3d dip, auto const &p) {
-        return dip + dipole_moment(p, box);
-      });
+                        combined_ranges const &p_q_unfolded_pos_range) {
+  auto const local_dip =
+      boost::accumulate(p_q_unfolded_pos_range, Utils::Vector3d{},
+                        [](Utils::Vector3d dip, auto const &q_pos) {
+                          double p_q{};
+                          Utils::Vector3d p_unfolded_pos{};
+                          boost::tie(p_q, p_unfolded_pos) = q_pos;
```

Comment:
Please double-check if this compiles, and if this improves the generated code. The idea would be to skip the copy of argument `dip`, skip the zero-initialization of `p_unfolded_pos`, and skip the copy of `p_unfolded_pos`.
```suggestion
                        [](Utils::Vector3d const &dip, auto const &q_pos) {
                          auto const p_q = boost::get<0>(zipped);
                          auto const &p_unfolded_pos = boost::get<1>(zipped);
```